README.md
# Microbenchmark

|master|
|--|
| [![Build Status](https://travis-ci.com/rai-project/microbench.svg?branch=master)](https://travis-ci.com/rai-project/microbench)|

A benchmark framework developed by the [IMPACT](impact.crhc.illinois.edu) group at the University of Illinois in collaboration with IBM through the the Center for Cognitive Computing Systems Research (C3SR).

Primary maintainers:
* Carl Pearson ([web](cwpearson.github.io)) ([email](mailto:pearson@illinois.edu))
* Abdul Dakkak ([email](mailto:dakkak@illinois.edu))
* Cheng Li ([email](mailto:cli99@illinois.edu))

Various benchmark suites using Scope are under development:

* [Comm|Scope](github.com/rai-project/comm_scope) - CUDA/NUMA data transfer performance
* [NCCL|Scope](github.com/rai-project/nccl_scope) - GPU collective communication performance
* [Misc|Scope](github.com/rai-project/misc_scope) - experimental or miscellaneous benchmarks


## Install CMake >= 3.12

On **x86-64**, the following may work.

```
cd /tmp
wget https://cmake.org/files/v3.12/cmake-3.12.0-Linux-x86_64.sh
sudo sh cmake-3.12.0-Linux-x86_64.sh --prefix=/usr/local --exclude-subdir
```

On **ppc64le**, you will need to build from source.

## Checkout all submodules

```
git submodule update --init --recursive
```

or to update modules, if you did not do a recursive clone

```
git submodule update --recursive --remote
```

## Compile

To compile the project run the following commands

    mkdir -p build
    cd build
    cmake -DCMAKE_BUILD_TYPE=Release ..
    make
    
The build system uses Hunter to download all dependencies.
If you have trouble downloading dependencies, [check to make sure](docs/hunter_problems.md) Hunter/CMake can use SSL.
Or you can forego Hunter entirely and [provide your own dependencies](docs/build_without_hunter.md).

if you get errors about nvcc not supporting your gcc compiler, then you may want to use

    cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_CUDA_HOST_COMPILER=`which gcc-6` .. 

You can optionally choose your own CUDA archs that you would like to be compiled:

    cmake -DNVCC_ARCH_FLAGS="2.0 2.1 2.0 2.1 3.0 3.2 3.5 3.7 5.0 5.2 5.3" ..

The accepted syntax is the same as the `CUDA_SELECT_NVCC_ARCH_FLAGS` syntax in the FindCUDA module.

You can disable or enable individual scopes

    cmake -DENABLE_MISC=0 ...

## Available Benchmarks

The available benchmarks and descriptions are listed [here](docs/benchmark_descriptions.md). You can list all the benchmarks with

    ./scope --benchmark_list_tests=true

you can filter the benchmarks that are run with a regular expression passed to `--benchmark_filter`.

    ./scope --benchmark_filter=[regex]

for example

    ./scope --benchmark_filter=SGEMM

futher controls over the benchmarks are explained in the `--help` option

## Run all the benchmarks

This is not generally recommended, as it will take quite some time.

    ./scope

The above will output to stdout something like 

    ------------------------------------------------------------------------------
    Benchmark                       Time           CPU Iterations UserCounters...
    ------------------------------------------------------------------------------
    SGEMM/1000/1/1/-1/1             5 us          5 us     126475 K=1 M=1000 N=1 alpha=-1 beta=1
    SGEMM/128/169/1728/1/0        539 us        534 us       1314 K=1.728k M=128 N=169 alpha=1 beta=0
    SGEMM/128/729/1200/1/0       1042 us       1035 us        689 K=1.2k M=128 N=729 alpha=1 beta=0
    SGEMM/192/169/1728/1/0        729 us        724 us        869 K=1.728k M=192 N=169 alpha=1 beta=0
    SGEMM/256/169/1/1/1             9 us          9 us      75928 K=1 M=256 N=169 alpha=1 beta=1
    SGEMM/256/729/1/1/1            35 us         35 us      20285 K=1 M=256 N=729 alpha=1 beta=1
    SGEMM/384/169/1/1/1            18 us         18 us      45886 K=1 M=384 N=169 alpha=1 beta=1
    SGEMM/384/169/2304/1/0       2475 us       2412 us        327 K=2.304k M=384 N=169 alpha=1 beta=0
    SGEMM/50/1000/1/1/1            10 us         10 us      73312 K=1 M=50 N=1000 alpha=1 beta=1
    SGEMM/50/1000/4096/1/0       6364 us       5803 us        100 K=4.096k M=50 N=1000 alpha=1 beta=0
    SGEMM/50/4096/1/1/1            46 us         45 us      13491 K=1 M=50 N=4.096k alpha=1 beta=1
    SGEMM/50/4096/4096/1/0      29223 us      26913 us         20 K=4.096k M=50 N=4.096k alpha=1 beta=0
    SGEMM/50/4096/9216/1/0      55410 us      55181 us         10 K=9.216k M=50 N=4.096k alpha=1 beta=0
    SGEMM/96/3025/1/1/1            55 us         51 us      14408 K=1 M=96 N=3.025k alpha=1 beta=1
    SGEMM/96/3025/363/1/0        1313 us       1295 us        570 K=363 M=96 N=3.025k alpha=1 beta=0

Output as JSON using

    ./scope --benchmark_out_format=json --benchmark_out=test.json
    
or preferably 

    ./scope --benchmark_out_format=json --benchmark_out=`hostname`.json

Repeat benchmark runs with

    ./scope --benchmark_repetitions=5

## Plot Benchmark JSON files

Try the [ScopePlot](https://github.com/rai-project/scope_plot) python package.

    pip install scope_plot

## On Minsky With PowerAI

```
cd build && rm -fr * && OpenBLAS=/opt/DL/openblas cmake -DCMAKE_BUILD_TYPE=Release .. -DOpenBLAS=/opt/DL/openblas
```

## Disable CPU frequency scaling

If you see this error:

```
***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
```

you might want to disable the CPU frequency scaling while running the benchmark:

```bash
sudo cpupower frequency-set --governor performance
./mybench
sudo cpupower frequency-set --governor powersave
```

## Run with Docker

Install `nvidia-docker`, then, list the available benchmarks.

    nvidia-docker run  --rm raiproject/microbench:amd64-latest bench --benchmark_list_tests

You can run benchmarks in the following way (probably with the `--benchmark_filter` flag).

    nvidia-docker run --privileged --rm -v `readlink -f .`:/data -u `id -u`:`id -g` raiproject/microbench:amd64-latest ./numa-separate-process.sh dgx bench /data/sync2

* `--privileged` is needed to set the NUMA policy if NUMA benchmarks are to be run.
* `` -v `readlink -f .`:/data `` maps the current directory into the container as `/data`.
* `` --benchmark_out=/data/\`hostname`.json `` tells the `bench` binary to write the json output files to `/data` in the container, which is mapped to the current directory.
* `` -u `id -u`:`id -g` `` tells docker to run as user `id -u` and group `id -g`, which is the current user and group. This means that files that docker produces will be modifiable from the host system without root permission.

## Hunter Toolchain File

If some of the third-party code compiled by hunter needs a different compiler, you can create a cmake toolchain file to set various cmake variables that will be globally used when building that code. You can then pass this file into cmake

    cmake -DCMAKE_TOOLCHAIN_FILE=toolchain.cmake ...

## Adding a new benchmark

If you would like to extend the benchmark suite, read [here](docs/new_benchmark.md) for more information.

## Third-Party Resources

* https://github.com/google/benchmark


CHANGELOG.md
Scope 0.1.0
-----------
* First versioned release
* Comm|Scope [764e8a5](https://github.com/rai-project/comm_scope/tree/764e8a5349d0d8fa30977b8490979fc14a85c7c3)
* Example|Scope [73789e7](https://github.com/rai-project/example_scope/tree/73789e7192dd141dfa7e9f50d7dfa498faa901bd)
* Misc|Scope  [73feb09](https://github.com/rai-project/misc_scope/tree/73feb09cc674a83299c83bf39077dfaae242d8f1)
* NCCL|Scope [cd2127a](https://github.com/rai-project/nccl_scope/tree/cd2127a68e4574d5c5918f4ae2a1fd6035d6eeb2)
* Logging and CUDA error handling utilities

src/sgemm.cpp
#include <benchmark/benchmark.h>

#include <cblas.h>
#include <iostream>
#include <numeric>
#include <stdio.h>
#include <stdlib.h>
#include <vector>

static void SGEMM(benchmark::State &state) {
  const auto M = state.range(0);
  const auto N = state.range(1);
  const auto K = state.range(2);
  const auto alpha = 1.0f;
  const auto beta = 0.0f;

  for (auto _ : state) {
    state.PauseTiming();
    auto a = std::vector<float>(M * K);
    auto b = std::vector<float>(K * N);
    auto c = std::vector<float>(M * N);
    std::iota(a.begin(), a.end(), 1);
    std::iota(b.begin(), b.end(), 1);
    std::fill(c.begin(), c.end(), 0);
    state.ResumeTiming();

    cblas_sgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, M, N, K, alpha,
                a.data(), K, b.data(), N, beta, c.data(), N);
  }

  state.counters.insert({{"M", M}, {"N", N}, {"K", K}});
}

BENCHMARK(SGEMM) // M, N, K
    ->Args({1000, 1, 1})
    ->Args({128, 169, 1728})
    ->Args({128, 729, 1200})
    ->Args({192, 169, 1728})
    ->Args({256, 169, 1})
    ->Args({256, 729, 1})
    ->Args({384, 169, 1})
    ->Args({384, 169, 2304})
    ->Args({50, 1000, 1})
    ->Args({50, 1000, 4096})
    ->Args({50, 4096, 1})
    ->Args({50, 4096, 4096})
    ->Args({50, 4096, 9216})
    ->Args({96, 3025, 1})
    ->Args({96, 3025, 363})
    ->Args({5124, 700, 2048})
    ->Args({35, 700, 2048})
    ->Args({5124, 700, 2560})
    ->Args({35, 700, 2560})
    ->Args({5124, 1500, 2048})
    ->Args({35, 1500, 2048})
    ->Args({5124, 1500, 2560})
    ->Args({35, 1500, 2560})
    ->Args({7680, 1, 2560})
    ->Args({7680, 2, 2560})
    ->Args({7680, 4, 2560})
    ->Args({3072, 1, 1024})
    ->Args({3072, 2, 1024})
    ->Args({3072, 4, 1024})
    ->Args({512, 1, 500000})
    ->Args({1024, 1, 500000})
    ->Args({512, 2, 500000})
    ->Args({1024, 2, 500000})
    ->Args({512, 4, 500000})
    ->Args({1024, 4, 500000})
    ->Args({1024, 700, 512})
    ->Args({7680, 1500, 2560})
    ->Args({6144, 1500, 2048})
    ->Args({4608, 1500, 1536})
    ->Args({8448, 1500, 2816})
    ->Args({3072, 1500, 1024})
    ->Args({7680, 3000, 2560})
    ->Args({6144, 3000, 2048})
    ->Args({4608, 3000, 1536})
    ->Args({8448, 3000, 2816})
    ->Args({3072, 3000, 1024})
    ->Args({7680, 6000, 2560})
    ->Args({6144, 6000, 2048})
    ->Args({4608, 6000, 1536})
    ->Args({8448, 6000, 2816})
    ->Args({3072, 6000, 1024})
    ->Args({6144, 1, 2048})
    ->Args({4608, 1, 1536})
    ->Args({8448, 1, 2816})
    ->Args({6144, 2, 2048})
    ->Args({4608, 2, 1536})
    ->Args({8448, 2, 2816})
    ->Args({6144, 4, 2048})
    ->Args({4608, 4, 1536})
    ->Args({8448, 4, 2816})
    ->Args({512, 1500, 2816})
    ->Args({512, 1500, 2048})
    ->Args({512, 1500, 2560})
    ->Args({512, 1500, 1536})
    ->Args({1024, 1500, 2816})
    ->Args({1024, 1500, 2048})
    ->Args({1024, 1500, 2560})
    ->Args({1024, 1500, 1536})
    ->Args({512, 1, 512})
    ->Args({1024, 1, 512})
    ->Args({512, 3000, 2816})
    ->Args({512, 3000, 2048})
    ->Args({512, 3000, 2560})
    ->Args({512, 3000, 1536})
    ->Args({1024, 3000, 2816})
    ->Args({1024, 3000, 2048})
    ->Args({1024, 3000, 2560})
    ->Args({1024, 3000, 1536})
    ->Args({512, 2, 512})
    ->Args({1024, 2, 512})
    ->Args({512, 6000, 2816})
    ->Args({512, 6000, 2048})
    ->Args({512, 6000, 2560})
    ->Args({512, 6000, 1536})
    ->Args({1024, 6000, 2816})
    ->Args({1024, 6000, 2048})
    ->Args({1024, 6000, 2560})
    ->Args({1024, 6000, 1536})
    ->Args({512, 4, 512})
    ->Args({1024, 4, 512})
    ->Unit(benchmark::kNanosecond);
